File: app/src/main/java/com/example/offlineassistant/CommandParser.kt
package com.example.offlineassistant

data class CommandResult(
    val action: String,
    val targetApp: String,
    val payload: String = "",
    val response: String,
    val cleanResponse: String
)

object CommandParser {
    
    private fun isAppOpenIntent(input: String, appName: String): Boolean {
        val keywords = listOf("open", "launch", "start", "run", "go to", "show")
        val hasKeyword = keywords.any { input.contains(it) }
        if (!hasKeyword) return false
        
        return when (appName) {
            "youtube" -> input.contains("youtube") || input.contains("yt")
            "chrome" -> input.contains("chrome") || input.contains("browser") || input.contains("internet")
            "whatsapp" -> input.contains("whatsapp")
            "gallery" -> input.contains("gallery") || input.contains("photo") || input.contains("image") || input.contains("photos")
            "maps" -> input.contains("maps") || input.contains("map")
            "settings" -> input.contains("settings") || input.contains("setting")
            else -> input.contains(appName)
        }
    }

    fun parse(rawText: String, activeApp: String): CommandResult {
        // Normalize spacing variations
        val text = rawText.lowercase().trim()
            .replace(Regex("you\\s+tube"), "youtube")
            .replace(Regex("whats\\s+app"), "whatsapp")
            .replace(Regex("what\\s+app"), "whatsapp")
            .replace(Regex("wi\\s+fi"), "wifi")
            .replace(Regex("wi-\\s*fi"), "wifi")
            .replace(Regex("blue\\s+tooth"), "bluetooth")
            .replace(Regex("google\\s+maps"), "maps")
            .replace(Regex("google\\s+chrome"), "chrome")

        // 1. SYSTEM CONTROLS
        if (text.contains("increase volume") || text.contains("volume up")) {
            return CommandResult("VOLUME_UP", "system", "", "Volume increased", "Volume increased")
        }
        if (text.contains("decrease volume") || text.contains("volume down")) {
            return CommandResult("VOLUME_DOWN", "system", "", "Volume decreased", "Volume decreased")
        }
        if (text.contains("increase brightness") || text.contains("brightness up")) {
            return CommandResult("BRIGHTNESS_UP", "system", "", "Brightness increased", "Brightness increased")
        }
        if (text.contains("decrease brightness") || text.contains("brightness down")) {
            return CommandResult("BRIGHTNESS_DOWN", "system", "", "Brightness decreased", "Brightness decreased")
        }
        if (text == "open wifi settings" || text.contains("wifi settings")) {
            return CommandResult("OPEN_WIFI", "settings", "", "Opening Wi-Fi settings", "Opening Wi-Fi settings")
        }
        if (text == "open bluetooth settings" || text.contains("bluetooth settings")) {
            return CommandResult("OPEN_BLUETOOTH", "settings", "", "Opening Bluetooth settings", "Opening Bluetooth settings")
        }

        // 2. APP LAUNCHERS
        if (isAppOpenIntent(text, "youtube")) {
            return CommandResult("OPEN_APP", "youtube", "com.google.android.youtube", "YouTube is open", "YouTube is open")
        }
        if (isAppOpenIntent(text, "chrome")) {
            return CommandResult("OPEN_APP", "chrome", "com.android.chrome", "Chrome is open", "Chrome is open")
        }
        if (isAppOpenIntent(text, "whatsapp")) {
            return CommandResult("OPEN_APP", "whatsapp", "com.whatsapp", "WhatsApp is open", "WhatsApp is open")
        }
        if (isAppOpenIntent(text, "gallery")) {
            return CommandResult("OPEN_APP", "gallery", "com.google.android.apps.photos", "Gallery is open", "Gallery is open")
        }
        if (isAppOpenIntent(text, "maps")) {
            return CommandResult("OPEN_APP", "maps", "com.google.android.apps.maps", "Maps is open", "Maps is open")
        }
        if (isAppOpenIntent(text, "settings")) {
            return CommandResult("OPEN_APP", "settings", "com.android.settings", "Settings is open", "Settings is open")
        }

        // 3. APP SPECIFIC ACTIONS

        // YOUTUBE
        if (activeApp.contains("youtube") || text.contains("youtube")) {
            if (text.startsWith("search ")) {
                val query = rawText.substring(7).trim()
                return CommandResult("YT_SEARCH", "youtube", query, "Searching YouTube for $query", "Searching YouTube for $query")
            }
            if (text.contains("play the first video") || text.contains("play video") || text.contains("play first")) {
                return CommandResult("YT_PLAY_FIRST", "youtube", "", "Playing the first video", "Playing the first video")
            }
            if (text == "pause" || text == "pause video" || text == "stop video" || text == "play") {
                return CommandResult("YT_PAUSE", "youtube", "", "Done", "Done")
            }
            if (text == "scroll down") {
                return CommandResult("SCROLL_DOWN", "youtube", "", "Scrolling down", "Scrolling down")
            }
        }

        // CHROME
        if (activeApp.contains("chrome") || text.contains("chrome")) {
            if (text.startsWith("search google for ")) {
                val query = rawText.substring(18).trim()
                return CommandResult("CHROME_SEARCH", "chrome", query, "Searching Google for $query", "Searching Google for $query")
            }
            if (text.contains("open this website") || text.contains("open website")) {
                return CommandResult("CHROME_OPEN_SITE", "chrome", "", "Opening the website", "Opening the website")
            }
            if (text == "go back" || text == "back") {
                return CommandResult("CHROME_BACK", "chrome", "", "Going back", "Going back")
            }
        }

        // WHATSAPP
        if (activeApp.contains("whatsapp") || text.contains("whatsapp")) {
            if (text.contains("open a chat") || text.contains("open chat")) {
                return CommandResult("WA_OPEN_CHAT", "whatsapp", "", "Opening chat", "Opening chat")
            }
            if (text.startsWith("type ")) {
                val msg = rawText.substring(5).trim()
                return CommandResult("WA_TYPE", "whatsapp", msg, "Typed $msg", "Typed $msg")
            }
            if (text == "send the message" || text == "send message" || text == "send") {
                return CommandResult("WA_SEND", "whatsapp", "", "Message sent", "Message sent")
            }
        }

        // GALLERY
        if (activeApp.contains("gallery") || activeApp.contains("photos") || text.contains("gallery")) {
            if (text.contains("open the latest photo") || text.contains("open latest photo") || text.contains("open photo")) {
                return CommandResult("GALLERY_OPEN_LATEST", "gallery", "", "Opening latest photo", "Opening latest photo")
            }
            if (text == "scroll down") {
                return CommandResult("SCROLL_DOWN", "gallery", "", "Scrolling down", "Scrolling down")
            }
        }

        // MAPS
        if (activeApp.contains("maps") || text.contains("maps")) {
            if (text.startsWith("search ")) {
                val dest = rawText.substring(7).trim()
                return CommandResult("MAPS_SEARCH", "maps", dest, "Searching for $dest on Maps", "Searching for $dest on Maps")
            }
            if (text.contains("start navigation") || text.contains("navigate")) {
                return CommandResult("MAPS_NAVIGATE", "maps", "", "Starting navigation", "Starting navigation")
            }
        }

        if (text == "go home" || text == "open home" || text == "close app") {
            return CommandResult("GO_HOME", "home", "", "Returned to Home screen", "Returned to Home screen")
        }

        return CommandResult("UNKNOWN", "system", "", "I didn't quite catch that", "I didnt quite catch that")
    }
}

File: src/utils/CommandParser.js
/**
 * Offline Command Parser Utility
 * Parses spoken or typed commands into structured actions that the simulator can execute.
 * It also returns the clean verbal response that will be spoken via offline Text-to-Speech (TTS).
 */

// Helper to match app open intents (e.g. "open youtube", "open the youtube", "launch youtube")
const isAppOpenIntent = (input, appName) => {
  const keywords = ["open", "launch", "start", "run", "go to", "show"];
  const hasKeyword = keywords.some(kw => input.includes(kw));
  if (!hasKeyword) return false;

  switch (appName) {
    case "youtube":
      return input.includes("youtube") || input.includes("yt");
    case "chrome":
      return input.includes("chrome") || input.includes("browser");
    case "whatsapp":
      return input.includes("whatsapp");
    case "gallery":
      return input.includes("gallery") || input.includes("photo") || input.includes("image");
    case "maps":
      return input.includes("maps") || input.includes("map");
    case "settings":
      return input.includes("settings") || input.includes("setting");
    default:
      return input.includes(appName);
  }
};

export const parseCommand = (rawText, permissions, activeApp) => {
  // Normalize spoken/typed app names with spacing variations
  const text = rawText.toLowerCase().trim()
    .replace(/\byou\s+tube\b/g, "youtube")
    .replace(/\bwhats\s+app\b/g, "whatsapp")
    .replace(/\bwhat\s+app\b/g, "whatsapp")
    .replace(/\bwi\s+fi\b/g, "wifi")
    .replace(/\bwi-\s*fi\b/g, "wifi")
    .replace(/\bblue\s+tooth\b/g, "bluetooth")
    .replace(/\bgoogle\s+maps\b/g, "maps")
    .replace(/\bgoogle\s+chrome\b/g, "chrome");

  // Helper to check if an app has permission
  const checkPermission = (appName) => {
    if (!permissions.accessibility) {
      return {
        error: true,
        response: "I need Accessibility permission enabled to control apps."
      };
    }
    if (!permissions.apps[appName]) {
      return {
        error: true,
        response: `Control for ${appName} is disabled in your assistant permissions.`
      };
    }
    return { error: false };
  };

  // 1. SYSTEM CONTROLS (Always available if global settings allow, or mic permission enabled)
  if (text.includes("increase volume") || text.includes("volume up")) {
    return {
      action: "VOLUME_UP",
      app: "system",
      response: "Volume increased",
      speechText: "Volume increased"
    };
  }

  if (text.includes("decrease volume") || text.includes("volume down")) {
    return {
      action: "VOLUME_DOWN",
      app: "system",
      response: "Volume decreased",
      speechText: "Volume decreased"
    };
  }

  if (text.includes("increase brightness") || text.includes("brightness up")) {
    return {
      action: "BRIGHTNESS_UP",
      app: "system",
      response: "Brightness increased",
      speechText: "Brightness increased"
    };
  }

  if (text.includes("decrease brightness") || text.includes("brightness down")) {
    return {
      action: "BRIGHTNESS_DOWN",
      app: "system",
      response: "Brightness decreased",
      speechText: "Brightness decreased"
    };
  }

  if (text === "open wifi settings" || text.includes("wifi settings")) {
    const perm = checkPermission("settings");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_WIFI",
      app: "settings",
      response: "Opening Wi-Fi settings",
      speechText: "Opening Wi-Fi settings"
    };
  }

  if (text === "open bluetooth settings" || text.includes("bluetooth settings")) {
    const perm = checkPermission("settings");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_BLUETOOTH",
      app: "settings",
      response: "Opening Bluetooth settings",
      speechText: "Opening Bluetooth settings"
    };
  }

  if (isAppOpenIntent(text, "settings")) {
    const perm = checkPermission("settings");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_APP",
      app: "settings",
      response: "Settings is open",
      speechText: "Settings is open"
    };
  }

  // 2. APP LAUNCHERS
  if (isAppOpenIntent(text, "youtube")) {
    const perm = checkPermission("youtube");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_APP",
      app: "youtube",
      response: "YouTube is open",
      speechText: "YouTube is open"
    };
  }

  if (isAppOpenIntent(text, "chrome")) {
    const perm = checkPermission("chrome");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_APP",
      app: "chrome",
      response: "Chrome is open",
      speechText: "Chrome is open"
    };
  }

  if (isAppOpenIntent(text, "whatsapp")) {
    const perm = checkPermission("whatsapp");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_APP",
      app: "whatsapp",
      response: "WhatsApp is open",
      speechText: "WhatsApp is open"
    };
  }

  if (isAppOpenIntent(text, "gallery")) {
    const perm = checkPermission("gallery");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_APP",
      app: "gallery",
      response: "Gallery is open",
      speechText: "Gallery is open"
    };
  }

  if (isAppOpenIntent(text, "maps")) {
    const perm = checkPermission("maps");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };
    return {
      action: "OPEN_APP",
      app: "maps",
      response: "Maps is open",
      speechText: "Maps is open"
    };
  }

  // 3. APP SPECIFIC ACTIONS (Accessibility triggers)
  
  // YOUTUBE
  if (activeApp === "youtube" || text.includes("youtube")) {
    const perm = checkPermission("youtube");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };

    if (text.startsWith("search ")) {
      const query = rawText.substring(7).trim(); // Extract query case-sensitively
      return {
        action: "YT_SEARCH",
        app: "youtube",
        payload: query,
        response: `Searching YouTube for "${query}"`,
        speechText: `Searching YouTube for ${query}`
      };
    }
    if (text.includes("play the first video") || text.includes("play video")) {
      return {
        action: "YT_PLAY_FIRST",
        app: "youtube",
        response: "Playing the first video",
        speechText: "Playing the first video"
      };
    }
    if (text === "pause" || text === "pause video" || text === "stop video") {
      return {
        action: "YT_PAUSE",
        app: "youtube",
        response: "Video paused",
        speechText: "Video paused"
      };
    }
    if (text === "scroll down") {
      return {
        action: "SCROLL_DOWN",
        app: "youtube",
        response: "Scrolling down",
        speechText: "Scrolling down"
      };
    }
  }

  // CHROME
  if (activeApp === "chrome" || text.includes("chrome")) {
    const perm = checkPermission("chrome");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };

    if (text.startsWith("search google for ")) {
      const query = rawText.substring(18).trim();
      return {
        action: "CHROME_SEARCH",
        app: "chrome",
        payload: query,
        response: `Searching Google for "${query}"`,
        speechText: `Searching Google for ${query}`
      };
    }
    if (text.includes("open this website") || text.includes("open website")) {
      return {
        action: "CHROME_OPEN_SITE",
        app: "chrome",
        response: "Opening the website",
        speechText: "Opening the website"
      };
    }
    if (text === "go back" || text === "back") {
      return {
        action: "CHROME_BACK",
        app: "chrome",
        response: "Going back",
        speechText: "Going back"
      };
    }
  }

  // WHATSAPP
  if (activeApp === "whatsapp" || text.includes("whatsapp")) {
    const perm = checkPermission("whatsapp");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };

    if (text.includes("open a chat") || text.includes("open chat")) {
      return {
        action: "WA_OPEN_CHAT",
        app: "whatsapp",
        response: "Opening chat with John",
        speechText: "Opening chat with John"
      };
    }
    if (text.startsWith("type ")) {
      const msg = rawText.substring(5).trim();
      return {
        action: "WA_TYPE",
        app: "whatsapp",
        payload: msg,
        response: `Typed "${msg}"`,
        speechText: `Typed ${msg}`
      };
    }
    if (text === "send the message" || text === "send message" || text === "send") {
      return {
        action: "WA_SEND",
        app: "whatsapp",
        response: "Message sent",
        speechText: "Message sent"
      };
    }
  }

  // GALLERY
  if (activeApp === "gallery" || text.includes("gallery")) {
    const perm = checkPermission("gallery");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };

    if (text.includes("open the latest photo") || text.includes("open latest photo") || text.includes("open photo")) {
      return {
        action: "GALLERY_OPEN_LATEST",
        app: "gallery",
        response: "Opening latest photo",
        speechText: "Opening latest photo"
      };
    }
    if (text === "scroll down") {
      return {
        action: "SCROLL_DOWN",
        app: "gallery",
        response: "Scrolling down Gallery",
        speechText: "Scrolling down Gallery"
      };
    }
  }

  // MAPS
  if (activeApp === "maps" || text.includes("maps")) {
    const perm = checkPermission("maps");
    if (perm.error) return { action: "PERMISSION_DENIED", app: "system", ...perm };

    if (text.startsWith("search ")) {
      const destination = rawText.substring(7).trim();
      return {
        action: "MAPS_SEARCH",
        app: "maps",
        payload: destination,
        response: `Searching for "${destination}" on Maps`,
        speechText: `Searching for ${destination} on Maps`
      };
    }
    if (text.includes("start navigation") || text.includes("navigate")) {
      return {
        action: "MAPS_NAVIGATE",
        app: "maps",
        response: "Starting navigation",
        speechText: "Starting navigation"
      };
    }
  }

  // HOME SCREEN (Go back to home)
  if (text === "go home" || text === "open home" || text === "close app") {
    return {
      action: "GO_HOME",
      app: "home",
      response: "Returned to Home screen",
      speechText: "Returned to Home screen"
    };
  }

  // Default / Command not understood
  return {
    action: "UNKNOWN",
    app: "system",
    response: "I didn't quite catch that. Try saying 'Open YouTube' or 'Increase volume'.",
    speechText: "I didn't quite catch that"
  };
};

/**
 * Clean voice output text for Speech Synthesis.
 * Strips punctuation to optimize raw text speech flow as requested:
 * e.g., "Done! Volume increased." -> "Done Volume increased"
 */
export const cleanSpeechResponse = (text) => {
  if (!text) return "";
  return text
    .replace(/[!,./;:?]/g, "") // remove punctuation
    .replace(/\s+/g, " ")       // normalize spacing
    .trim();
};
